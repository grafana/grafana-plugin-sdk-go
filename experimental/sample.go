package experimental

import (
	"fmt"
	"os"

	"github.com/grafana/grafana-plugin-sdk-go/backend"
)

//----------------------------------------------------------------------------------
// PLUGIN HOST
//----------------------------------------------------------------------------------

// MyHost is a singleton host service.
type MyHost struct{}

// CheckHostHealth returns a backend.CheckHealthResult.
func (ds *MyHost) CheckHostHealth(ctx backend.PluginContext) *backend.CheckHealthResult {
	return &backend.CheckHealthResult{
		Status:  backend.HealthStatusOk,
		Message: "Plugin is running",
	}
}

// NewDataSourceInstance creates a new datasource instance.
func (ds *MyHost) NewDataSourceInstance(ctx backend.PluginContext) (DataSourceInstance, error) {
	settings := myDataSourceSettings{
		url:  ctx.DataSourceInstanceSettings.URL,
		port: 1234,
	}

	return &MyDataSourceInstance{
		settings: settings,
		cache:    "TODO",
		client:   "TODO",
	}, nil
}

//----------------------------------------------------------------------------------
// DATA SOURCE instance
//----------------------------------------------------------------------------------

type myDataSourceSettings struct {
	url  string
	port int32
}

// MyDataSourceInstance implements backend.DataSourceInstance
type MyDataSourceInstance struct {
	settings myDataSourceSettings
	cache    interface{} // for example
	client   interface{} // for example
}

// CheckHealth will check the currently configured settings
func (ds *MyDataSourceInstance) CheckHealth() *backend.CheckHealthResult {
	if len(ds.settings.url) < 2 {
		return &backend.CheckHealthResult{
			Status:  backend.HealthStatusError,
			Message: "invalid URL",
		}
	}
	return &backend.CheckHealthResult{
		Status:  backend.HealthStatusOk,
		Message: "Datasource is setup properly",
	}
}

// QueryData will run a set of queries
func (ds *MyDataSourceInstance) QueryData(req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
	return nil, fmt.Errorf("not implemented")
}

// CallResource HTTP style resource
func (ds *MyDataSourceInstance) CallResource(req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {

	if req.Path == "hello" {
		return SendJSON(sender, map[string]interface{}{"hello": "world"})
	}

	if req.Path == "text" {
		return SendPlainText(sender, "hello world")
	}

	return fmt.Errorf("unknown resource")
}

// Destroy destroy an instance (if necessary)
func (ds *MyDataSourceInstance) Destroy() {
	// If necessary, destroy the object (typically not required)
}

//----------------------------------------------------------------------------------
// SAMPLE MAIN
//----------------------------------------------------------------------------------

// MainXYZ is a sample main function.
func MainXYZ() {
	// Setup the plugin environment
	backend.SetupPluginEnvironment("my-datasource")

	backend.Logger.Debug("Running my backend datasource")

	host := NewInstanceManager(&MyHost{})
	err := host.RunGRPCServer()
	if err != nil {
		backend.Logger.Error(err.Error())
		os.Exit(1)
	}
}
