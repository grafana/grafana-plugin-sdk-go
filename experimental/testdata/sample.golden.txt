Frame[0] {
    "executedQueryString": "SELECT * FROM X",
    "notices": [
        {
            "text": "hello"
        }
    ]
}
Name: Frame One
Dimensions: 1 Fields by 5 Rows
+----------------------+
| Name: Single float64 |
| Labels:              |
| Type: []float64      |
+----------------------+
| 8.26                 |
| 8.7                  |
| 14.82                |
| 10.07                |
| 8.52                 |
+----------------------+


Frame[1] 
Name: Frame Two
Dimensions: 1 Fields by 3 Rows
+---------------------+
| Name: single string |
| Labels: a=b         |
| Type: []string      |
+---------------------+
| a                   |
| b                   |
| c                   |
+---------------------+



====== TEST DATA RESPONSE (arrow base64) ======
FRAME=QVJST1cxAAD/////wAEAABAAAAAAAAoADgAMAAsABAAKAAAAFAAAAAAAAAEDAAoADAAAAAgABAAKAAAACAAAAMAAAAADAAAAVAAAACgAAAAEAAAA6P7//wgAAAAMAAAAAAAAAAAAAAAFAAAAcmVmSWQAAAAI////CAAAABQAAAAJAAAARnJhbWUgT25lAAAABAAAAG5hbWUAAAAAMP///wgAAABQAAAARgAAAHsiZXhlY3V0ZWRRdWVyeVN0cmluZyI6IlNFTEVDVCAqIEZST00gWCIsIm5vdGljZXMiOlt7InRleHQiOiJoZWxsbyJ9XX0AAAQAAABtZXRhAAAAAAEAAAAYAAAAAAASABgAFAAAABMADAAAAAgABAASAAAAFAAAAIAAAACIAAAAAAAAA4gAAAACAAAAPAAAAAQAAADU////CAAAABgAAAAOAAAAU2luZ2xlIGZsb2F0NjQAAAQAAABuYW1lAAAAAAgADAAIAAQACAAAAAgAAAAcAAAAEgAAAHsidW5pdCI6IlBlcmNlbnQifQAABgAAAGNvbmZpZwAAAAAAAAAABgAIAAYABgAAAAAAAgAOAAAAU2luZ2xlIGZsb2F0NjQAAAAAAAD/////iAAAABQAAAAAAAAADAAWABQAEwAMAAQADAAAACgAAAAAAAAAFAAAAAAAAAMDAAoAGAAMAAgABAAKAAAAFAAAADgAAAAFAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAKAAAAAAAAAAAAAAAAQAAAAUAAAAAAAAAAAAAAAAAAACF61G4HoUgQGZmZmZmZiFApHA9CtejLUCkcD0K1yMkQArXo3A9CiFAEAAAAAwAFAASAAwACAAEAAwAAAAQAAAALAAAADgAAAAAAAMAAQAAANABAAAAAAAAkAAAAAAAAAAoAAAAAAAAAAAAAAAAAAAAAAAKAAwAAAAIAAQACgAAAAgAAADAAAAAAwAAAFQAAAAoAAAABAAAAOj+//8IAAAADAAAAAAAAAAAAAAABQAAAHJlZklkAAAACP///wgAAAAUAAAACQAAAEZyYW1lIE9uZQAAAAQAAABuYW1lAAAAADD///8IAAAAUAAAAEYAAAB7ImV4ZWN1dGVkUXVlcnlTdHJpbmciOiJTRUxFQ1QgKiBGUk9NIFgiLCJub3RpY2VzIjpbeyJ0ZXh0IjoiaGVsbG8ifV19AAAEAAAAbWV0YQAAAAABAAAAGAAAAAAAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAACAAAAAiAAAAAAAAAOIAAAAAgAAADwAAAAEAAAA1P///wgAAAAYAAAADgAAAFNpbmdsZSBmbG9hdDY0AAAEAAAAbmFtZQAAAAAIAAwACAAEAAgAAAAIAAAAHAAAABIAAAB7InVuaXQiOiJQZXJjZW50In0AAAYAAABjb25maWcAAAAAAAAAAAYACAAGAAYAAAAAAAIADgAAAFNpbmdsZSBmbG9hdDY0AADoAQAAQVJST1cx
FRAME=QVJST1cxAAD/////gAEAABAAAAAAAAoADgAMAAsABAAKAAAAFAAAAAAAAAEDAAoADAAAAAgABAAKAAAACAAAAFgAAAACAAAAKAAAAAQAAAAg////CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAAED///8IAAAAFAAAAAkAAABGcmFtZSBUd28AAAAEAAAAbmFtZQAAAAABAAAAGAAAAAAAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAACwAAAAtAAAAAAAAAWwAAAAAwAAAGgAAAA0AAAABAAAAKz///8IAAAAGAAAAA0AAABzaW5nbGUgc3RyaW5nAAAABAAAAG5hbWUAAAAA2P///wgAAAAUAAAACQAAAHsiYSI6ImIifQAAAAYAAABsYWJlbHMAAAgADAAIAAQACAAAAAgAAAAgAAAAFQAAAHsiZGlzcGxheU5hbWUiOiIxMjMifQAAAAYAAABjb25maWcAAAAAAAAEAAQABAAAAA0AAABzaW5nbGUgc3RyaW5nAAAAAAAAAP////+YAAAAFAAAAAAAAAAMABYAFAATAAwABAAMAAAAGAAAAAAAAAAUAAAAAAAAAwMACgAYAAwACAAEAAoAAAAUAAAASAAAAAMAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAABAAAAAAAAAACAAAAAAAAAAAAAAAAQAAAAMAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAIAAAADAAAAYWJjAAAAAAAQAAAADAAUABIADAAIAAQADAAAABAAAAAsAAAAOAAAAAAAAwABAAAAkAEAAAAAAACgAAAAAAAAABgAAAAAAAAAAAAAAAAAAAAAAAoADAAAAAgABAAKAAAACAAAAFgAAAACAAAAKAAAAAQAAAAg////CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAAED///8IAAAAFAAAAAkAAABGcmFtZSBUd28AAAAEAAAAbmFtZQAAAAABAAAAGAAAAAAAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAACwAAAAtAAAAAAAAAWwAAAAAwAAAGgAAAA0AAAABAAAAKz///8IAAAAGAAAAA0AAABzaW5nbGUgc3RyaW5nAAAABAAAAG5hbWUAAAAA2P///wgAAAAUAAAACQAAAHsiYSI6ImIifQAAAAYAAABsYWJlbHMAAAgADAAIAAQACAAAAAgAAAAgAAAAFQAAAHsiZGlzcGxheU5hbWUiOiIxMjMifQAAAAYAAABjb25maWcAAAAAAAAEAAQABAAAAA0AAABzaW5nbGUgc3RyaW5nAAAAqAEAAEFSUk9XMQ==
